File: sdk/typescript/src/cost_calculator.ts
/**
 * Cost calculator for the Axon TypeScript SDK.
 *
 * Uses BigInt arithmetic scaled to 8 decimal places for all monetary
 * calculations. No floating-point arithmetic is used for currency values
 * (ADR-006 equivalent for TypeScript).
 */

import { PRICING } from "./pricing";

/**
 * Convert a price-per-1M-tokens string to a BigInt in units of 1e-8 USD.
 *
 * Internally the result represents (price_usd_per_1M * 1e8) as a BigInt,
 * so that multiplying by a token count and dividing by 1_000_000 yields
 * total cost in units of 1e-8 USD without any floating-point operations.
 */
function priceStringToUnits(priceStr: string): bigint {
  // We allow at most 8 decimal places in the price string.
  const parts = priceStr.split(".");
  const intPart = parts[0] ?? "0";
  const fracPart = (parts[1] ?? "").padEnd(8, "0").slice(0, 8);
  return BigInt(intPart) * BigInt("100000000") + BigInt(fracPart);
}

/**
 * Calculate the USD cost of an LLM inference call using string-based
 * fixed-point arithmetic.
 *
 * No floating-point arithmetic is used for monetary values. The result is
 * a string with exactly 8 decimal places (e.g. `"0.00250000"`).
 *
 * @param model - The model identifier (e.g. `"gpt-4o"`).
 * @param inputTokens - Number of input/prompt tokens.
 * @param outputTokens - Number of output/completion tokens.
 * @param cachedTokens - Tokens served from provider cache; billed at the
 *   `cacheReadCostPer1MTokens` rate. Defaults to `0`.
 * @returns Cost string with 8 decimal places, or `null` for unknown models.
 */
export function calculateCost(
  model: string,
  inputTokens: number,
  outputTokens: number,
  cachedTokens = 0,
): string | null {
  const pricing = PRICING[model];
  if (pricing === undefined) {
    return null;
  }

  if (inputTokens === 0 && outputTokens === 0 && cachedTokens === 0) {
    return "0.00000000";
  }

  // Scale factor: 1 USD = 100_000_000 units (1e8)
  const ONE_MILLION = BigInt("1000000");
  const SCALE = BigInt("100000000"); // 1e8

  const inputRate = priceStringToUnits(pricing.inputCostPer1MTokens);
  const outputRate = priceStringToUnits(pricing.outputCostPer1MTokens);

  // Regular (non-cached) input tokens
  const regularInput = BigInt(Math.max(0, inputTokens - cachedTokens));
  let total = (regularInput * inputRate) / ONE_MILLION;
  total += (BigInt(outputTokens) * outputRate) / ONE_MILLION;

  // Cached tokens billed at cache-read rate (falls back to input rate)
  if (cachedTokens > 0) {
    const cacheRate =
      pricing.cacheReadCostPer1MTokens !== undefined
        ? priceStringToUnits(pricing.cacheReadCostPer1MTokens)
        : inputRate;
    total += (BigInt(cachedTokens) * cacheRate) / ONE_MILLION;
  }

  // Format: total is in units of 1e-8 USD
  const whole = total / SCALE;
  const frac = total % SCALE;
  return `${whole.toString()}.${frac.toString().padStart(8, "0")}`;
}

File: sdk/python/traject/core/cost_calculator.py
"""Cost calculation utilities for LLM inference calls.

Provides ``get_pricing`` for model price lookups and ``calculate_cost`` for
computing the exact USD cost of an inference call using Decimal arithmetic
throughout (ADR-006) — no float values are used at any point in the
computation pipeline.
"""

from __future__ import annotations

from decimal import Decimal

import structlog

from traject.core.pricing import PROVIDER_PRICING, ModelPricing

_log = structlog.get_logger(__name__)


def get_pricing(model: str) -> ModelPricing | None:
    """Return pricing data for the given model identifier, or ``None``.

    Performs a direct key lookup against the static ``PROVIDER_PRICING``
    table.  Unknown model strings are silently ignored — no exception is
    raised and no warning is logged.

    Args:
        model: The model identifier string as used in provider API requests
            (e.g. ``"gpt-4o"``).

    Returns:
        A :class:`~traject.core.pricing.ModelPricing` instance if the model is
        present in the pricing table, or ``None`` if it is unknown.
    """
    return PROVIDER_PRICING.get(model)


def calculate_cost(
    model: str,
    input_tokens: int,
    output_tokens: int,
    cached_tokens: int = 0,
) -> Decimal | None:
    """Compute the total USD cost for a single inference call.

    All arithmetic is performed exclusively with :class:`~decimal.Decimal`
    values (ADR-006).  No ``float`` values are introduced at any point.  The
    result is quantized to eight decimal places.

    If the model is not present in the pricing table, a ``WARNING``-level
    structured log event is emitted and ``None`` is returned — no exception
    is raised.

    Args:
        model: The model identifier string (e.g. ``"gpt-4o"``).
        input_tokens: Total number of input (prompt) tokens billed, including
            any tokens that were served from the provider's prompt cache.
        output_tokens: Total number of output (completion) tokens billed.
        cached_tokens: Number of input tokens that were served from the
            provider's prompt cache and should therefore be billed at the
            cache-read rate rather than the standard input rate.  Defaults to
            ``0``.  Must be ``<= input_tokens``.

    Returns:
        The total cost as a :class:`~decimal.Decimal` quantized to eight
        decimal places (e.g. ``Decimal("0.00250000")``), or ``None`` if the
        model is not found in the pricing table.

    Notes:
        The formula applied is::

            MILLION = Decimal("1000000")
            non_cached_input = input_tokens - cached_tokens
            input_cost  = (Decimal(non_cached_input) / MILLION)
                          * pricing.input_cost_per_1m_tokens
            cache_cost  = (Decimal(cached_tokens)    / MILLION)
                          * pricing.cache_read_cost_per_1m_tokens
                          (only when cached_tokens > 0 and cache rate is not None)
            output_cost = (Decimal(output_tokens)    / MILLION)
                          * pricing.output_cost_per_1m_tokens
            total = (input_cost + cache_cost + output_cost)
                    .quantize(Decimal("0.00000001"))
    """
    pricing = get_pricing(model)
    if pricing is None:
        _log.warning("traject.cost.unknown_model", model=model)
        return None

    million = Decimal("1000000")

    non_cached_input = input_tokens - cached_tokens
    input_cost = (
        Decimal(non_cached_input) / million
    ) * pricing.input_cost_per_1m_tokens

    cache_cost = Decimal("0")
    if cached_tokens > 0 and pricing.cache_read_cost_per_1m_tokens is not None:
        cache_cost = (
            Decimal(cached_tokens) / million
        ) * pricing.cache_read_cost_per_1m_tokens

    output_cost = (Decimal(output_tokens) / million) * pricing.output_cost_per_1m_tokens

    total = input_cost + cache_cost + output_cost
    return total.quantize(Decimal("0.00000001"))
